soh/soh/Enhancements/stairs.cpp

```diff
@@ -0,0 +1,675 @@
+    { ACTOR_EN_GS,                 0x1F00},
+    { ACTOR_BG_MIZU_BWALL,         0x17B0},
+    { ACTOR_BG_MIZU_SHUTTER,       0x890},
+    { ACTOR_EN_DAIKU_KAKARIKO,     0x14B0},
+    { ACTOR_BG_BOWL_WALL,          0xA80},
+    { ACTOR_EN_WALL_TUBO,          0x730},
+    { ACTOR_EN_PO_DESERT,          0xE50},
+    { ACTOR_EN_CROW,               0x1670},
+    { ACTOR_DOOR_KILLER,           0x1650},
+    { ACTOR_BG_SPOT11_OASIS,       0x7A0},
+    { ACTOR_BG_SPOT18_FUTA,        0x190},
+    { ACTOR_BG_SPOT18_SHUTTER,     0x540},
+    { ACTOR_EN_MA3,                0x1060},
+    { ACTOR_EN_COW,                0x1440},
+    { ACTOR_BG_ICE_TURARA,         0x820},
+    { ACTOR_BG_ICE_SHUTTER,        0x460},
+    { ACTOR_EN_KAKASI2,            0xD10},
+    { ACTOR_EN_KAKASI3,            0x1280},
+    { ACTOR_OCEFF_WIPE4,           0x10B0},
+    { ACTOR_EN_EG,                 0x210},
+    { ACTOR_BG_MENKURI_NISEKABE,   0x150},
+    { ACTOR_EN_ZO,                 0x27A0},
+    { ACTOR_OBJ_MAKEKINSUTA,       0x2A0},
+    { ACTOR_EN_GE3,                0xCF0},
+    { ACTOR_OBJ_TIMEBLOCK,         0xDC0},
+    { ACTOR_OBJ_HAMISHI,           0x8D0},
+    { ACTOR_EN_ZL4,                0x4A70},
+    { ACTOR_EN_MM2,                0xED0},
+    { ACTOR_BG_JYA_BLOCK,          0x300},
+    { ACTOR_OBJ_WARP2BLOCK,        0xCB0},
+};
+
+std::unordered_set<u16> registeredStairOverlays;
+
+std::unordered_map<uintptr_t, uintptr_t> generalPtrMap;
+std::unordered_map<u16, uintptr_t> overlayRPtrMap;
+std::unordered_map<u16, uintptr_t> overlayPtrMap;
+std::unordered_map<u16, uintptr_t> mallocRPtrMap;
+
+void* absolutePtr = nullptr;
+
+s32 absoluteSpaceFlag = 0;
+
+u16 Stairs_GetAllocType(u16 id) {
+    for (const ActorIdAllocTypePair& pair : stairActorTable) {
+        if (pair.id == id) {
+            return pair.allocType;
+        }
+    }
+    return 0xFFFF;
+}
+
+s32 Stairs_RegisterOverlay(u16 id) {
+    if (registeredStairOverlays.contains(id)) {
+        return false;
+    }
+    registeredStairOverlays.insert(id);
+    return true;
+}
+
+s32 Stairs_GetOverlayRegistered(u16 id) {
+    return registeredStairOverlays.contains(id);
+}
+
+s32 Stairs_UnregisterOverlay(u16 id) {
+    if (registeredStairOverlays.contains(id)) {
+        registeredStairOverlays.erase(id);
+        return true;
+    }
+    return false;
+}
+
+s32 Stairs_RegisterAbsoluteSpace() {
+    if (absoluteSpaceFlag) {
+        return false;
+    }
+    absoluteSpaceFlag = true;
+    return true;
+}
+
+s32 Stairs_GetAbsoluteSpace() {
+    return absoluteSpaceFlag;
+}
+
+void Stairs_UnregisterAbsoluteSpace() {
+    absoluteSpaceFlag = false;
+}
+
+size_t Stairs_GetOverlaySize(u16 id) {
+    return actorOverlaySizes[id];
+}
+
+size_t Stairs_GetSize() {
+    return stairsAllocSize;
+}
+
+s32 Stairs_DecreaseSize(size_t size) {
+    if (size > stairsAllocSize) {
+        return false;
+    }
+    stairsAllocSize -= size;
+    return true;
+}
+
+void* StairsArena_Malloc(size_t size) {
```

Comment:
This might just be lack of knowledge on my part, but I don't really understand what the difference is between `Malloc`, `MallocR`, `MallocGeneral`, `MallocRGeneral`, etc. are. If that's the kind of thing that can be briefly explained in a comment somewhere in here I think that would be good, if not I'll ask you at some point later because it's probably just some deep memory management knowledge that I'm missing.